### ISSUE TITLE
Instance feedback for Soundcraft Ui Series

### ISSUE THREAD

**Issue by DachsbauTV** [bitfocus/companion-module-soundcraft-ui#3]
Hi everyone,

It would be great if there would be a feedback instance for at least the muting state of the channels.

numeric feedback for fader level would be a plus

**Comment by fmalcher** [comment]
Can somebody show a good example implementation of how we can set instance feedback?

**Comment by DachsbauTV** [comment]
Seems like these guys have made feedback work on the quite similar Behringer x32 

https://github.com/bitfocus/companion-module-behringer-x32

Hope this helps  🙌

**Comment by fmalcher** [comment]
Good to know, thanks! In the meantime, I have worked on a separate library that allows communicating with the Soundcraft mixer. It also handles the mixer state and resolves feedback. The enxt step will be to integrate the library into this module. After this, we can actually bring the feedback to Companion.

**Comment by DachsbauTV** [comment]
That's amazing! you won't believe how much easier my life will be thanks to that! Thanks already

**Comment by JeffreyDavidsz** [comment]
@fmalcher i run into issues, according to the api sending a INIT command should return a JSON object. It doesn't it returns text lines. They say keep the connection alive each second by sending ALIVE. From that you should get updated json objects. Because that all doesn't work, I tried to resend the INIT command several times, but after a while no data will return. How did you keep receiving the data?

**Comment by fmalcher** [comment]
@JeffreyDavidsz I'm not really sure what API description you're talking about. can you please elaborate? 🙂
The mixer keeps sending incremental updates via se socket connection. I used those to assemble a full state object.

I'm currently working on integrating my connection lib into this companion module, see my draft PR. Feedback already works in theory, but I couldn't test it with the real hardware yet.

**Comment by JeffreyDavidsz** [comment]
I have an api from Soundcraft (not allowed to send) the initial status is coming in, but after that nothing untill I send INIT. I'll check the pr!

**Comment by JeffreyDavidsz** [comment]
@fmalcher Besides the original module had it already, is there a reason why you use a random number in the connection string?
`ws://${this.targetIP}/socket.io/1/websocket/${Math.random() * 10000}` It might be the solution of my update problems, but I would like to know why

**Comment by fmalcher** [comment]
@JeffreyDavidsz Actually, I just took this from the original module. This might be due to how socket.io (which is used by the original Soundcraft web app) structures its URLs. Maybe @TheMrPoseidon can tell us more?

I wanted to use Socket.io as well, but the mixer has a super old version of the protocol that is not supported anymore by recent releases of socket.io. Thus, I still stuck with plain websocket through `rxjs/websocket`.

**Comment by JeffreyDavidsz** [comment]
@fmalcher I don't see any INIT command command in your code to keep receiving updates?

**Comment by JeffreyDavidsz** [comment]
> 
> 
> @JeffreyDavidsz Actually, I just took this from the original module. This might be due to how socket.io (which is used by the original Soundcraft web app) structures its URLs. Maybe @TheMrPoseidon can tell us more?
> 
> I wanted to use Socket.io as well, but the mixer has a super old version of the protocol that is not supported anymore by recent releases of socket.io. Thus, I still stuck with plain websocket through `rxjs/websocket`.

clear, I already spoke about it, I remember main thing was stability.
So my main issue is still how to receive incremental updates

**Comment by fmalcher** [comment]
> @fmalcher I don't see any INIT command command in your code to keep receiving updates?

You're right! The current implementation of this module doesn't use it either and it also worked for me without.
Does the original Soundcraft webclient send an `INIT` or is this just written in the API spec?

btw, you can ping me in the Slack at any time if you like to chat!

**Comment by TheMrPoseidon** [comment]
@JeffreyDavidsz @fmalcher The reason of why I used a random number was that I had issues reconnecting with the mixer. I assumed that the number behind `/websocket/` was some sort of a session id. When I allocated a random number after the connection got lost everything worked fine. I tested `ws://${this.targetIP}/` with the new connection handling from @fmalcher. I see no problem in changing it.
